### PR TITLE
fix(pnpify): make go to definition work with coc.nvim

### PR DESCRIPTION
**Pull request by cometkim**
**What's the problem this PR addresses?**

fixes https://github.com/neoclide/coc-tsserver/issues/244

https://twitter.com/KrComet/status/1370033304464855044

`gd`(Go to Definition) is not working for [coc-tsserver](https://github.com/neoclide/coc-tsserver) host.

...

**How did you fix it?**

I was looking for a way to search inside zip with vim, and it seems to possible with [vim-rzip](https://github.com/lbrayner/vim-rzip) plugin.

This patch makes it possible to resolve paths compatible with vim-rzip in the coc-nvim host.

https://github.com/lbrayner/vim-rzip is required for the user's vim config.

...

### CODE REVIEW

File: packages/yarnpkg-pnpify/sources/sdks/base.ts [hunk 1]
```diff
@@ -72,15 +72,29 @@ export const generateTypescriptBaseWrapper: GenerateBaseWrapper = async (pnpApi:
...
+              // Ref: https://github.com/microsoft/vscode/issues/105014#issuecomment-686760910
+              //
+              case \`vscode\`:
+                str = \`^zip:\${str}\`; break;
+
+              // To make "go to definition work",
+              // resolve the actual file system path from virtual path
+              // and convert to scheme supported by [vim-rzip](https://github.com/lbrayner/vim-rzip)
+              case \`coc-nvim\`:
+                str = str.replace(/\\$\\$virtual\\/([\\-\\w\\/]+)\\/cache/, \`cache\`);
```
Comment (merceyz) [inline packages/yarnpkg-pnpify/sources/sdks/base.ts:91]: The `resolved` variable contains the real path https://github.com/yarnpkg/berry/blob/b27cd5ebf40b2dbe80e81b7b3ba77b0edfe28ecb/packages/yarnpkg-pnpify/sources/sdks/base.ts#L64

Comment (cometkim) [reply packages/yarnpkg-pnpify/sources/sdks/base.ts]: it works, thanks!

**Comment by arcanis** [comment]
I'm not sure what but something seems wrong - running `yarn pnpify --sdk` in the repo then refreshing the VSCode window stops the TS server from spawning 🤔

**Comment by cometkim** [comment]
> I'm not sure what but something seems wrong - running `yarn pnpify --sdk` in the repo then refreshing the VSCode window stops the TS server from spawning 🤔

Hmm... still there is some syntax error..? I will check

**Comment by timtyrrell** [comment]
I will drop this here in case it is of any value with the PR. Seemed like this could work without that extra vim plugin? https://github.com/neoclide/coc-tsserver/issues/244#issuecomment-755271565

**Comment by cometkim** [comment]
@timtyrrell Yes it is the same scheme so it seems to work. the user-friendly way is to guide the plugin (as we're doing with vscode), but it seems worth mentioning for users who don't want it.

**Comment by cometkim** [comment]
@arcanis I fixed a bug, can you test the last commit?

**Comment by cometkim** [comment]
I ran pnpify here and checked it in action.

Please checkout the new commit and let me know if go to definition doesn't work properly.

**Comment by cometkim** [comment]
Does this change contain anything that could affect the Windows environment? I don't know why it fails :thinking:

**Comment by merceyz** [comment]
You're not normalizing `str` anymore

**Comment by cometkim** [comment]
@merceyz Ah you right

File: .yarn/sdks/typescript/lib/tsserver.js [hunk 2]
```diff
@@ -34,26 +37,39 @@ const moduleWrapper = tsserver => {
...
       if (resolved) {
         const locator = pnpApi.findPackageLocator(resolved);
         if (locator && dependencyTreeRoots.has(`${locator.name}@${locator.reference}`)) {
-         str = resolved;
+          str = resolved;
         }
       }
 
-      str = str.replace(/\\/g, `/`)
-      str = str.replace(/^\/?/, `/`);
```
Comment (merceyz) [inline .yarn/sdks/typescript/lib/tsserver.js:46]: These still need to happen for `default` and `vscode`

**Review by merceyz — CHANGES_REQUESTED** [review]

File: .yarn/sdks/typescript/lib/tsserver.js (hunk 2 above)
Comment (cometkim) [reply .yarn/sdks/typescript/lib/tsserver.js:46]: yep I just fixed it

**Comment by arcanis** [comment]
Perfect, just tried again and no issue with VSCode anymore. Thanks @cometkim ! 😃

**Pull request merged**